Init-system misdetection on Linux hosts in a mixed Linux+Windows cluster (SysVinit selected instead of Systemd)

**Module / version:** `github.com/k0sproject/rig/v2` — reproduced on `v2.0.0`, `v2.0.1`, `v2.1.0`

## Summary

In a process that manages **both Linux and Windows hosts**, rig can resolve the wrong
`ServiceManager` for a systemd-based Linux host — selecting **SysVinit** instead of
**Systemd**. `SysVinit.ServiceIsRunning` then probes a nonexistent `/etc/init.d/<svc>`
script, so `Service.IsRunning()` returns `false` for a service that is actually running, and
`Start/Stop/Restart/Enable/Disable` target the wrong init system.

Root cause is the interaction of three things:

1. `initsystem.DefaultRegistry()` is a **process-wide singleton** whose factory slice is
   shared across every host.
2. `plumbing.Provider.Get` returns the first matching factory **and moves it to the front of
   the shared slice** ("optimize future lookups", assuming a homogeneous cluster).
3. The **SysVinit factory matches systemd hosts** — its probe `test -d /etc/init.d` succeeds
   on virtually every modern systemd distro (the directory is kept for LSB compatibility).

Systemd wins the Systemd-vs-SysVinit tie **only** because it is registered first. Once a
Windows host resolves (moving `WinSCM` to the front, demoting `Systemd`), a subsequently
resolved Linux host matches `SysVinit` before it ever reaches the demoted `Systemd`.

## Minimal reproduction (unit test, no hosts/runners)

The defect reduces to a single property of `plumbing.Provider`: the move-to-front reorder
changes which of two *both-matching* factories wins. This needs nothing but the exported
`plumbing` API — no runners, no remote commands, no `initsystem` package:

```go
package plumbing_test

import (
	"errors"
	"testing"

	"github.com/k0sproject/rig/v2/plumbing"
)

func TestProviderMoveToFrontBreaksMultiMatchPrecedence(t *testing.T) {
	// Model of the real registry: "systemd" and "sysvinit" BOTH match a Linux host
	// (on Ubuntu, `stat /run/systemd/system` and `test -d /etc/init.d` both succeed);
	// "winscm" matches a Windows host. systemd is registered first, so it should
	// always win for Linux — the tie is broken by registration order alone.
	newRegistry := func() *plumbing.Provider[string, string] {
		p := plumbing.NewProvider[string, string](errors.New("no match"))
		p.Register(func(h string) (string, bool) { return "systemd", h == "linux" })  // idx 0
		p.Register(func(h string) (string, bool) { return "sysvinit", h == "linux" }) // idx 1, also matches linux
		p.Register(func(h string) (string, bool) { return "winscm", h == "windows" }) // idx 2
		return p
	}

	// Homogeneous cluster: a Linux host always resolves to systemd.
	p := newRegistry()
	if got, _ := p.Get("linux"); got != "systemd" {
		t.Fatalf("linux-only: got %q, want systemd", got)
	}

	// Mixed cluster: a Windows host resolves FIRST — matching "winscm" moves it to the
	// front of the shared slice, demoting "systemd".
	p = newRegistry()
	p.Get("windows")

	// The same Linux host now resolves to sysvinit, because systemd is no longer tried first.
	if got, _ := p.Get("linux"); got != "systemd" {
		t.Fatalf("BUG: after a Windows host resolved, a Linux host resolved to %q; want systemd", got)
	}
}
```

Result on `v2.0.0`, `v2.0.1`, and `v2.1.0`:

```
--- FAIL: TestProviderMoveToFrontBreaksMultiMatchPrecedence
    provider_test.go: BUG: after a Windows host resolved, a Linux host resolved to "sysvinit"; want systemd
```

## Why it happens (the real registry)

`Client.Service()` eagerly resolves the manager and caches it per connection:

```go
// client.go
func (c *Client) Service(name string) (*Service, error) {
    is, err := c.ServiceManager()   // -> initsystem.DefaultRegistry().Get(runner), cached via LazyService
    ...
}
```

`DefaultRegistry()` is a shared singleton registering factories in fixed order
(`initsystem/defaultprovider.go`):

```
Systemd, OpenRC, Upstart, SysVinit, WinSCM, Runit, Launchd
```

`plumbing.Provider.Get` (`plumbing/provider.go`):

```go
func (p *Provider[R, T]) Get(r R) (T, error) {
    p.mu.Lock(); defer p.mu.Unlock()
    for i, f := range p.factories {
        t, ok := f(r)
        if ok {
            if i != 0 {
                // moved to front "to optimize future lookups, since it's likely
                // that most of the hosts ... will be running the same kind of environment."
                p.factories[0], p.factories[i] = p.factories[i], p.factories[0]
            }
            return t, nil
        }
    }
    return *new(T), p.err
}
```

Detection probes that both match an Ubuntu 22.04 host:

```go
// RegisterSystemd:  stat /run/systemd/system   -> matches
// RegisterSysVinit: test -d /etc/init.d        -> ALSO matches (LSB compat dir)
```

Walkthrough:

1. Slice starts `[Systemd, OpenRC, Upstart, SysVinit, WinSCM, Runit, Launchd]`.
2. A **Windows** host resolves → `WinSCM` matches at index 4 → swapped to front →
   `[WinSCM, OpenRC, Upstart, SysVinit, Systemd, Runit, Launchd]`. **Systemd demoted.**
3. A **Linux** host resolves → WinSCM/OpenRC/Upstart miss → **SysVinit matches at index 3**,
   before the demoted Systemd → the Linux host gets `SysVinit`.
4. `SysVinit.ServiceIsRunning(ctx, h, "<systemd-unit>")` runs `/etc/init.d/<unit> status`,
   which does not exist → returns `false`. No error surfaces (SysVinit is a *valid* manager),
   so callers see a running service reported as stopped.

### Why a Linux-only cluster is unaffected, but one Windows node breaks it

Because `Systemd` is registered first (index 0), and in a Linux-only cluster nothing ever
displaces it from that position — so it is always the first factory tried, always matches an
Ubuntu host, and short-circuits the loop before `SysVinit` (index 3) is ever reached.

The detail is in `plumbing.Provider.Get`: it iterates the factory slice in order, returns the
first match, and only reorders when the match was **not** already at the front (`if i != 0`).

**Linux-only cluster** — slice stays in registration order:

```
[Systemd, OpenRC, Upstart, SysVinit, WinSCM, Runit, Launchd]
```

The first Linux host resolves → `Systemd` matches at index 0 → returned immediately; `i == 0`,
so no reorder. `SysVinit` at index 3 is never even evaluated. Every subsequent Linux host hits
the same index-0 `Systemd`. `SysVinit` can't win because it is never reached.

Both factories *do* match an Ubuntu host (`stat /run/systemd/system` and `test -d /etc/init.d`
both succeed) — but registration order is the tie-breaker, and `Systemd` is registered before
`SysVinit`. As long as `Systemd` holds index 0, that precedence holds.

**What breaks it (Windows present)** — a Windows host is the only thing that matches a factory
sitting *behind* `Systemd`, and thereby triggers a reorder:

```
Windows host → WinSCM matches at index 4 → i != 0 → swapped to front:
[WinSCM, OpenRC, Upstart, SysVinit, Systemd, Runit, Launchd]   ← Systemd demoted to index 4
```

Now a Linux host walks the reordered slice: WinSCM (no), OpenRC (no), Upstart (no), `SysVinit`
at index 3 matches → returned — before it ever reaches the demoted `Systemd` at index 4.

**One-line answer:** without a Windows host, no factory that outranks-by-position ever gets
displaced, so `Systemd` never loses index 0, and the move-to-front — which only relocates a
factory that actually matched — never fires in a way that promotes `SysVinit` ahead of
`Systemd`. It takes a Windows host matching a later factory (`WinSCM`) to push `Systemd` behind
`SysVinit` in the shared slice.

That's also exactly why the minimal repro calls `Get("windows")` before the second
`Get("linux")` — that single interleaved call is what demotes `Systemd` and flips the result.

## End-to-end reproduction (real registration functions, fake runners)

Confirms the same outcome through the actual `Register*` functions and `ServiceManager`
interface, with fake runners standing in for hosts:

```go
package main

import (
	"context"
	"fmt"
	"io"
	"strings"

	"github.com/k0sproject/rig/v2/cmd"
	"github.com/k0sproject/rig/v2/initsystem"
	"github.com/k0sproject/rig/v2/protocol"
)

type fakeRunner struct {
	name    string
	windows bool
	ok      map[string]bool // command substring -> exit-0
}

func (f *fakeRunner) String() string { return f.name }
func (f *fakeRunner) IsWindows() bool { return f.windows }
func (f *fakeRunner) ExecContext(_ context.Context, command string, _ ...cmd.ExecOption) error {
	for sub, good := range f.ok {
		if strings.Contains(command, sub) {
			if good {
				return nil
			}
			return fmt.Errorf("command failed: %s", command)
		}
	}
	return fmt.Errorf("not found: %s", command)
}
func (f *fakeRunner) ExecOutputContext(context.Context, string, ...cmd.ExecOption) (string, error) {
	return "", nil
}
func (f *fakeRunner) ExecReaderContext(context.Context, string, ...cmd.ExecOption) io.Reader { return nil }
func (f *fakeRunner) Start(context.Context, string, ...cmd.ExecOption) (protocol.Waiter, error) {
	return nil, nil
}

// Ubuntu 22.04: systemd present AND /etc/init.d present.
func ubuntu(name string) *fakeRunner {
	return &fakeRunner{name: name, ok: map[string]bool{
		"stat /run/systemd/system": true,
		"test -d /etc/init.d":      true,
	}}
}
func windows(name string) *fakeRunner { return &fakeRunner{name: name, windows: true} }

func freshRegistry() *initsystem.Registry {
	r := initsystem.NewRegistry()
	initsystem.RegisterSystemd(r)
	initsystem.RegisterOpenRC(r)
	initsystem.RegisterUpstart(r)
	initsystem.RegisterSysVinit(r)
	initsystem.RegisterWinSCM(r)
	initsystem.RegisterRunit(r)
	initsystem.RegisterLaunchd(r)
	return r
}

func name(reg *initsystem.Registry, r cmd.ContextRunner) string {
	sm, err := reg.Get(r)
	if err != nil {
		return "error: " + err.Error()
	}
	return sm.(fmt.Stringer).String()
}

func main() {
	a := freshRegistry()
	fmt.Println("Linux-only:")
	fmt.Println("  ubuntu ->", name(a, ubuntu("u1")))
	fmt.Println("  ubuntu ->", name(a, ubuntu("u2")))

	b := freshRegistry()
	fmt.Println("Windows resolves first (shared registry):")
	fmt.Println("  windows ->", name(b, windows("w1")))
	fmt.Println("  ubuntu  ->", name(b, ubuntu("u1")))
	fmt.Println("  ubuntu  ->", name(b, ubuntu("u2")))
}
```

Output:

```
Linux-only:
  ubuntu -> systemd
  ubuntu -> systemd
Windows resolves first (shared registry):
  windows -> winscm
  ubuntu  -> sysvinit      <-- should be systemd
  ubuntu  -> sysvinit
```

## Regression

This does not reproduce on the `v0.x` line (`github.com/k0sproject/rig`), which resolves the
init system per host with a fixed **systemd-first** check and no shared, reorderable registry
(`os/linux.go`):

```go
func (c Linux) is(h Host) (initSystem, error) {
	if c.hasSystemd(h) { // stat /run/systemd/system
		return &initsystem.Systemd{}, nil
	}
	if c.hasOpenRC(h) || c.hasUpstart(h) || c.hasSysV(h) {
		return &initsystem.OpenRC{}, nil
	}
	return nil, fmt.Errorf("%w: failed to detect init system", ErrInitSystemNotSupported)
}
```

`is(h)` is evaluated per host with no cross-host state, so a Windows host cannot influence a
Linux host's resolution, and systemd is always chosen when present.

| rig version | init-system resolution | minimal repro |
|---|---|---|
| `v0.21.x` | per-host `is(h)`, systemd-first, no shared registry | passes (mechanism absent) |
| `v2.0.0` | shared `DefaultRegistry` + move-to-front | **fails → sysvinit** |
| `v2.0.1` | same | **fails → sysvinit** |
| `v2.1.0` | same | **fails → sysvinit** |

So the behavior was introduced by the `v0.x → v2` rewrite — the shared registry, the
move-to-front reorder, and SysVinit matching systemd hosts — and is present in every `v2`
release to date.

## Impact

`Service.IsRunning()` returns `false` for a running systemd service on the Linux hosts of a
mixed Linux+Windows cluster, and `Start`/`Stop`/`Restart`/`Enable`/`Disable` likewise target
`/etc/init.d/<service>` and fail. It is deterministic whenever a Windows host resolves before a
Linux host, and reproduces on stock Ubuntu 22.04.

Any consumer that drives service operations across such a cluster is affected. For example,
k0sctl's reset gates a "stop the service before reset" step on `IsRunning()`; that stop is
silently skipped, and the subsequent `k0s reset` then aborts with:

```
Error: k0s seems to be running, please stop k0s before reset
```

## Possible fixes (not mutually exclusive)

1. **Don't let the reorder change the winner among multiple-matching factories.** The
   move-to-front optimization assumes cluster homogeneity and silently changes precedence.
   Options: drop the reorder; cache per-runner instead of mutating shared order; or give each
   `Client` its own registry rather than the shared singleton.
2. **Make SysVinit detection more specific** so it does not match systemd hosts — e.g. require
   the absence of systemd (`! stat /run/systemd/system`) in addition to `test -d /etc/init.d`,
   which alone is true on nearly every systemd distro.
3. **Order-independent detection:** score candidates by specificity and pick the most specific,
   rather than "first match wins after mutation."

Happy to open a PR for whichever direction you prefer.

--- a/initsystem/service_ops_test.go
+++ b/initsystem/service_ops_test.go
@@ -604,6 +604,13 @@ func TestRegistryDetection(t *testing.T) {
 
 		_, err := reg.Get(mr)
 		require.ErrorIs(t, err, initsystem.ErrNoInitSystem)
+
+		// Assert SysVinit actually exercised both probes: it passed the
+		// /etc/init.d check and then declined on the systemd-present check. This
+		// keeps the test specific so it cannot pass without running the intended
+		// detection logic.
+		require.NoError(t, mr.Received(rigtest.Contains("/etc/init.d")))
+		require.NoError(t, mr.Received(rigtest.Contains("stat /run/systemd/system")))
 	})
 
 	// Regression test for https://github.com/k0sproject/rig/issues/409: in a mixed

--- a/initsystem/service_ops_test.go
+++ b/initsystem/service_ops_test.go
@@ -590,6 +590,55 @@ func TestRegistryDetection(t *testing.T) {
 		require.ErrorIs(t, err, initsystem.ErrNoInitSystem)
 		require.NoError(t, mr.NotReceived(rigtest.Contains("openrc")))
 	})
+
+	t.Run("SysVinit not detected when systemd is present", func(t *testing.T) {
+		reg := initsystem.NewRegistry()
+		initsystem.RegisterSysVinit(reg)
+
+		// A systemd host: /etc/init.d exists (LSB compat) but systemd is present.
+		// SysVinit must decline so it never shadows systemd.
+		mr := rigtest.NewMockRunner()
+		mr.ErrDefault = errExec
+		mr.AddCommandSuccess(rigtest.Contains("/etc/init.d"))
+		mr.AddCommandSuccess(rigtest.Contains("stat /run/systemd/system"))
+
+		_, err := reg.Get(mr)
+		require.ErrorIs(t, err, initsystem.ErrNoInitSystem)
+	})
+
+	// Regression test for https://github.com/k0sproject/rig/issues/409: in a mixed
+	// Linux+Windows cluster, a Windows host resolving first swaps WinSCM to the front
+	// of the shared factory slice, demoting Systemd behind SysVinit. A subsequently
+	// resolved systemd Linux host must still resolve to Systemd, not SysVinit.
+	t.Run("Systemd wins on a systemd host after a Windows host resolved first (issue #409)", func(t *testing.T) {
+		reg := initsystem.NewRegistry()
+		// Same order as DefaultRegistry.
+		initsystem.RegisterSystemd(reg)
+		initsystem.RegisterOpenRC(reg)
+		initsystem.RegisterUpstart(reg)
+		initsystem.RegisterSysVinit(reg)
+		initsystem.RegisterWinSCM(reg)
+		initsystem.RegisterRunit(reg)
+		initsystem.RegisterLaunchd(reg)
+
+		// Windows host resolves first -> WinSCM matches and is moved to the front.
+		win := rigtest.NewMockRunner()
+		win.Windows = true
+		mgr, err := reg.Get(win)
+		require.NoError(t, err)
+		assert.IsType(t, &initsystem.WinSCM{}, mgr)
+
+		// Ubuntu-like systemd host: both `stat /run/systemd/system` and
+		// `test -d /etc/init.d` succeed. Must resolve to Systemd.
+		ubuntu := rigtest.NewMockRunner()
+		ubuntu.ErrDefault = errExec
+		ubuntu.AddCommandSuccess(rigtest.Contains("stat /run/systemd/system"))
+		ubuntu.AddCommandSuccess(rigtest.Contains("/etc/init.d"))
+
+		mgr, err = reg.Get(ubuntu)
+		require.NoError(t, err)
+		assert.IsType(t, initsystem.Systemd{}, mgr)
+	})
 }
 
 // Compile-time interface checks.

--- a/initsystem/sysvinit.go
+++ b/initsystem/sysvinit.go
@@ -87,11 +87,20 @@ func (i SysVinit) DisableService(ctx context.Context, h cmd.ContextRunner, s str
 
 // RegisterSysVinit registers SysVinit in a repository.
 func RegisterSysVinit(repo *Registry) {
-	repo.Register(func(c cmd.ContextRunner) (ServiceManager, bool) {
-		if c.IsWindows() {
+	repo.Register(func(runner cmd.ContextRunner) (ServiceManager, bool) {
+		if runner.IsWindows() {
 			return nil, false
 		}
-		if c.ExecContext(context.Background(), "test -d /etc/init.d") != nil {
+		if runner.ExecContext(context.Background(), "test -d /etc/init.d") != nil {
+			return nil, false
+		}
+		// /etc/init.d exists on virtually every systemd distro too (kept for LSB
+		// compatibility), so the probe above is not specific enough on its own: it
+		// makes SysVinit match systemd hosts. Require systemd to be ABSENT as well,
+		// mirroring RegisterSystemd's probe, so a systemd host never resolves to
+		// SysVinit regardless of factory ordering in the shared registry.
+		// See https://github.com/k0sproject/rig/issues/409.
+		if runner.ExecContext(context.Background(), "stat /run/systemd/system") == nil {
 			return nil, false
 		}
 

